### PR TITLE
Replace Scarb.toml tool config with snfoundry.toml 

### PR DESCRIPTION
**Pull request by THenry14**
https://github.com/foundry-rs/starknet-foundry/issues/1620
This PR is a part of stack, that aims to replace Scarb.toml tool configuration with snfoundry.toml:

-- Add mechanism for snfoundry.toml discovery (https://github.com/foundry-rs/starknet-foundry/pull/1643)
-> Replace Scarb.toml based configuration with snfoundry.toml based one
-- Use the same profile for sncast and scarb
-- Update docs

## Introduced changes

- use snfoundry.toml for tool configuration

## Checklist

- [X] Linked relevant issue
- [X] Updated relevant documentation
- [X] Added relevant tests
- [X] Performed self-review of the code
- [X] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by MaksymilianDemitraszek** [comment]
We probably need to update the changelog & docs.

**Comment by THenry14** [comment]
> We probably need to update the changelog & docs. Not sure if here is the good place to do so.

https://github.com/foundry-rs/starknet-foundry/pull/1647 , decided to do it in one place in separate PR, but if you'd rather have it here I can merge the changes

File: crates/sncast/src/helpers/configuration.rs [hunk 1]
```diff
@@ -1,7 +1,154 @@
...
+    fn default_value() -> Self {
+        Some(T::default_value())
+    }
+}
+
+pub fn get_profile(sncast_config: &Value, profile: &Option<String>) -> Result<Value> {
+    let profile_name = profile.as_deref().unwrap_or("default");
+    let config = sncast_config
+        .get("sncast")
+        .expect("Failed to find sncast config in foundry.toml file");
```
Comment (ddoktorski) [inline crates/sncast/src/helpers/configuration.rs:111]: nit: foundry.toml -> snfoundry.toml

File: crates/sncast/src/helpers/configuration.rs [hunk 2]
```diff
@@ -99,4 +246,49 @@ mod tests {
...
+            &Some(Utf8PathBuf::try_from(tempdir.path().to_path_buf()).unwrap()),
+        )
+        .unwrap();
+
+        assert_eq!(config.account, String::new());
+        assert_eq!(config.rpc_url, String::new());
+    }
+
+    #[test]
+    fn test_config_defaults() {
```
Comment (ddoktorski) [inline crates/sncast/src/helpers/configuration.rs:289]: nit: Maybe also add a test for `parse_config` when the path is `None`?

Comment (THenry14) [reply crates/sncast/src/helpers/configuration.rs:289]: This is basically impossible to test in unit tests :( I have experimented with sealed tests etc, but it all went down to the fact that we would need to set cwd for this test, and since the test env for all tests is the same, the other tests were executed in the new dir as well, and this resulted in many errors.

I do, however, test exactly this scenario in e2e tests, where we are able to set per-snapbox working directory, so it is generally covered :+1:

**Review by MaksymilianDemitraszek — COMMENTED** [review]
I looked only on the config loading part, looks good. 
Somebody with more cast specific domain knowledge should review this.

File: crates/sncast/src/helpers/configuration.rs
[the remark this answers is not in the document]
Comment (THenry14) [reply crates/sncast/src/helpers/configuration.rs:1]: This was to conform to current conventions and layouts. I am working right now (as a last part of this giga-stack) on a refactor to our main and lib to clean things up, and indeed I have a plan to move this (as well as constants) away from helpers dir, as these are not helpers

File: crates/sncast/src/helpers/configuration.rs [hunk 3]
```diff
@@ -1,7 +1,154 @@
...
+}
+
+pub fn get_property_optional<T>(entries: &Value, field: &str) -> Option<T>
+where
+    T: PropertyFromCastConfig + Default,
+{
+    entries.get(field).and_then(T::from_toml_value)
+}
+
+pub fn parse_config(profile: &Option<String>, path: &Option<Utf8PathBuf>) -> Result<CastConfig> {
```
[the remark this answers is not in the document]
Comment (THenry14) [reply crates/sncast/src/helpers/configuration.rs:134]: > I don't understand what path argument is here.

This can set a path to a directory we want to start searching from. Very useful for testing, but also for taking configs from specific packages (even if we are not in the package dir)

File: crates/sncast/src/starknet_commands/account/add.rs [hunk 4]
```diff
@@ -52,16 +51,16 @@ pub struct Add {
     #[clap(short, long)]
     pub salt: Option<FieldElement>,
 
-    /// If passed, a profile with corresponding data will be created in Scarb.toml
+    /// If passed, a profile with corresponding data will be created in snfoundry.toml
+    #[allow(clippy::struct_field_names)]
     #[clap(long)]
-    pub add_profile: bool,
+    pub add_profile: Option<String>,
```
[the remark this answers is not in the document]
Comment (THenry14) [reply crates/sncast/src/starknet_commands/account/add.rs:57]: Yeah, it's only partly related, could probably be moved, but it was easier for me to make other changes when this is a part of the change as well (it was hard to move to different PR).

Up until now, we have created new profiles with default name of an account (when this flag was passed). This changes this behaviour - now you can name you profile whatever you like, it is not tied to account name anymore.

File: crates/sncast/src/helpers/configuration.rs (hunk 1 above)
Comment (THenry14) [reply crates/sncast/src/helpers/configuration.rs]: Done in https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/de9f39d88ada985c9bcb12ea3676aeee67b85684

File: crates/sncast/src/helpers/configuration.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (THenry14) [reply crates/sncast/src/helpers/configuration.rs]: renamed in https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/de9f39d88ada985c9bcb12ea3676aeee67b85684

File: crates/sncast/src/helpers/configuration.rs [hunk 5]
```diff
@@ -99,4 +246,49 @@ mod tests {
             "Failed to find snfoundry.toml - not found in current nor any parent directories"
         );
     }
+
+    #[test]
+    fn parse_config_happy_case_with_profile() {
```
Comment (ddoktorski) [inline crates/sncast/src/helpers/configuration.rs:251]: nit: Since the name of the function has been changed, can we also rename these tests to `load_config...`?

File: crates/sncast/src/starknet_commands/account/add.rs [hunk 6]
```diff
@@ -52,16 +51,16 @@ pub struct Add {
     #[clap(short, long)]
     pub salt: Option<FieldElement>,
 
-    /// If passed, a profile with corresponding data will be created in Scarb.toml
+    /// If passed, a profile with corresponding data will be created in snfoundry.toml
```
Comment (ddoktorski) [inline crates/sncast/src/starknet_commands/account/add.rs:54]: nit:
```suggestion
    /// If passed, a profile with the provided name and corresponding data will be created in snfoundry.toml
```

File: crates/sncast/src/helpers/configuration.rs (hunk 5 above)
Comment (THenry14) [reply crates/sncast/src/helpers/configuration.rs]: done https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/321691a0c92cb4c5890734e9a931c1fd2c806607

File: crates/sncast/src/starknet_commands/account/add.rs (hunk 6 above)
Comment (THenry14) [reply crates/sncast/src/starknet_commands/account/add.rs]: changed (also in account/create) https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/321691a0c92cb4c5890734e9a931c1fd2c806607

File: crates/sncast/src/starknet_commands/account/add.rs [hunk 7]
```diff
@@ -92,21 +91,21 @@ pub async fn add(
...
         };
-        add_created_profile_to_configuration(path_to_scarb_toml, &config)?;
+        add_created_profile_to_configuration(&add.add_profile, &config, &None)?;
     }
 
     Ok(AccountAddResponse {
-        add_profile: if add.add_profile {
-            "Profile successfully added to Scarb.toml".to_string()
+        add_profile: if add.add_profile.is_some() {
+            "Profile successfully added to snfoundry.toml".to_string()
```
Comment (ddoktorski) [inline crates/sncast/src/starknet_commands/account/add.rs:106]: nit: It would be nice to include a name of the created profile here.

File: crates/sncast/src/starknet_commands/account/create.rs [hunk 8]
```diff
@@ -27,9 +27,9 @@ pub struct Create {
     #[clap(short, long)]
     pub salt: Option<FieldElement>,
 
-    /// If passed, a profile with corresponding data will be created in Scarb.toml
+    /// If passed, a profile with corresponding data will be created in snfoundry.toml
```
Comment (ddoktorski) [inline crates/sncast/src/starknet_commands/account/create.rs]: nit:
```suggestion
    /// If passed, a profile with the provided name and corresponding data will be created in snfoundry.toml
```

File: crates/sncast/src/starknet_commands/account/create.rs [hunk 9]
```diff
@@ -81,24 +80,24 @@ pub async fn create(
...
+        add_created_profile_to_configuration(&add_profile, &config, &None)?;
     }
 
     Ok(AccountCreateResponse {
         address: Hex(address),
         max_fee: Decimal(max_fee),
-        add_profile: if add_profile {
-            "Profile successfully added to Scarb.toml".to_string()
+        add_profile: if add_profile.is_some() {
+            "Profile successfully added to snfoundry.toml".to_string()
```
Comment (ddoktorski) [inline crates/sncast/src/starknet_commands/account/create.rs:98]: nit: The same comment as [here](https://github.com/foundry-rs/starknet-foundry/pull/1644/files#r1478354817)

File: crates/sncast/src/starknet_commands/account/add.rs (hunk 7 above)
Comment (THenry14) [reply crates/sncast/src/starknet_commands/account/add.rs]: added https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/af08cbf36a17154306ea6f8cad673856ebb4cb20

File: crates/sncast/src/starknet_commands/account/create.rs (hunk 9 above)
Comment (THenry14) [reply crates/sncast/src/starknet_commands/account/create.rs]: https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/af08cbf36a17154306ea6f8cad673856ebb4cb20

File: crates/sncast/tests/e2e/show_config.rs [hunk 10]
```diff
@@ -1,28 +1,22 @@
...
 #[tokio::test]
-async fn test_show_config_from_scarb_toml() {
-    let args = vec![
-        "--path-to-scarb-toml",
-        "tests/data/show_config/wait_Scarb.toml",
-        "--profile",
-        "profile1",
-        "show-config",
-    ];
+async fn test_show_config_from_scast_toml() {
```
Comment (ddoktorski) [inline crates/sncast/tests/e2e/show_config.rs:6]: nit:
```suggestion
async fn test_show_config_from_snfoundry_toml() {
```

File: crates/sncast/tests/e2e/show_config.rs [hunk 11]
```diff
@@ -56,74 +50,57 @@ async fn test_show_config_from_cli() {
...
-    let args = vec![
-        "--account",
-        "user2",
-        "--path-to-scarb-toml",
-        "tests/data/show_config/all_Scarb.toml",
-        "--profile",
-        "profile1",
-        "show-config",
-    ];
+async fn test_show_config_from_cli_and_sncast_toml() {
```
Comment (ddoktorski) [inline crates/sncast/tests/e2e/show_config.rs:53]: nit:
```suggestion
async fn test_show_config_from_cli_and_snfoundry_toml() {
```

**Review by ddoktorski — APPROVED** [review]
I've left a few more nitpick comments, but overall it looks good to me 💪

File: crates/sncast/tests/e2e/show_config.rs (hunk 10 above)
Comment (THenry14) [reply crates/sncast/tests/e2e/show_config.rs:6]: https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/1d8ff0cbbe7632b1bdb0a6a40092b5a5eabfccdb

File: crates/sncast/tests/e2e/show_config.rs (hunk 11 above)
Comment (THenry14) [reply crates/sncast/tests/e2e/show_config.rs]: https://github.com/foundry-rs/starknet-foundry/pull/1644/commits/1d8ff0cbbe7632b1bdb0a6a40092b5a5eabfccdb
